### ISSUE TITLE
Can you use Github Actions to build pre-compiled binary releases?

### ISSUE THREAD

**Issue by Rowisi** [Eloston/ungoogled-chromium#916]
Can you use `Github actions` to build and publish pre-compiled binary?

I am not experienced with Github actions but I can try and code it for you if you want.

I found an example for a repo written in GO: https://github.com/Jguer/yay/blob/master/.github/workflows/docker-multiarch-tag.yml

**Comment by wchen342** [comment]
I'm actually working on sth related, as a first step for solving #845. I'm testing it in my own repo now and once it worked I will make an Arch version too.
Just curious, is it because the binary version disappeared from AUR?

**Comment by Rowisi** [comment]
> Just curious, is it because the binary version disappeared from AUR?
No, I never used that. I have already made an OBS for myself and I can get binary package anytime.

I am speaking in general, For all users who wants to use this Browser including Windows, MacOS users, for x64 and ARM architectures, We need a pre-compiled binary releases attached for every update.

I think that Github Actions could be the solution for this.

**Comment by wchen342** [comment]
I see. Linux distributions are similar so it shouldn't be too hard to get them working once I can build the Arch version. Do not know about Windows and MacOS versions though.

**Comment by Rowisi** [comment]
> I see. Linux distributions are similar so it shouldn't be too hard to get them working once I can build the Arch version. Do not know about Windows and MacOS versions though.

We won't be making packages. We just need a compiled binary in a zip file and then people can make packages for their distributions based on it just like `brave browser`

For Windows we need to make an Installation exe with an integrated updater

For mac, we need to make  a .dmg file

**Comment by wchen342** [comment]
> We won't be making packages. We just need a compiled binary in a zip file and then people can make packages for their distributions based on it just like `brave browser`

There are three repos: `portablelinux`, `debian` and `archlinux`, so why not making all of them (binary, deb, pkg)?

> For Windows we need to make an Installation exe with an integrated updater
> 
> For mac, we need to make a .dmg file

I know what the file will be. I am saying I cannot test them because I do not have the system.

**Comment by Rowisi** [comment]
> There are three repos: portablelinux, debian and archlinux, so why not making all of them (binary, deb, pkg)?

Github Actions uses Ubuntu(Debian based) so we can only make binary and .deb

> I know what the file will be. I am saying I cannot test them because I do not have the system.

There are already `windows` and `macos`, we just have to build them.

**Comment by wchen342** [comment]
> Github Actions uses Ubuntu(Debian based) so we can only make binary and .deb

It's undocumented but Github Actions actually support Docker containers.

> There are already `windows` and `macos` repos, we just have to build them.

It will be hard to debug if anything goes wrong, without an actual system you can play with.

**Comment by tangalbert919** [comment]
> There are already `windows` and `macos` repos, we just have to build them.

I am currently working on the `windows` repo right now, and you can take a look [here](https://github.com/tangalbert919/ungoogled-chromium-windows/actions).

**Comment by Rowisi** [comment]
> I am currently working on the `windows` repo right now, and you can take a look [here](https://github.com/tangalbert919/ungoogled-chromium-windows/actions).


Good luck on that, I don't have free time now. Once I get free time, I'll try to make it for Linux first. First, I need to learn how to use Github actions.

**Comment by csagan5** [comment]
> It's undocumented but Github Actions actually support Docker containers.

You mean like this? https://help.github.com/en/actions/automating-your-workflow-with-github-actions/creating-a-docker-container-action

Would the GitHub-provided container have enough resources to build Chromium?

**Comment by wchen342** [comment]
@csagan5 What I used is actually [this](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#jobsjob_idcontainer), but it only works on Ubuntu VMs, which is kind of too slow. I have [an updated version running on macOS](https://github.com/wchen342/ungoogled-chromium-archlinux/runs/409417048) which is 25% faster.

**Comment by csagan5** [comment]
@wchen342 from what I understand it would not work for an Android build because it would take more than 6h?

I have considered in the past using https://build.opensuse.org/ (assuming they would allow it), but it would not be integrated with GitHub actions.

**Comment by wchen342** [comment]
@csagan5 It works for most single targets ([`MonoChrome+arm`](https://github.com/wchen342/ungoogled-chromium-android/actions/runs/31510904), [`Chrome+arm64`](https://github.com/wchen342/ungoogled-chromium-android/actions/runs/31510903)). The only exception is `MonoChrome+arm64` that has twice more ninja targets (which I have no idea why). I don't build `Trichrome` so I didn't try that.

**Comment by nikolowry** [comment]
@csagan5 I saw a few weeks ago that [Github Actions now supports self-hosted runners](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/about-self-hosted-runners).

I was planning on beginning automating builds with my beefiest private server (8gb ram, 120gb storage) before the months over. I'll report back my findings in this issue or another.

**Comment by csagan5** [comment]
> 8gb ram, 120gb storage

@nikolowry is that the one you mentioned takes ~4 hours to build? Is it a bare metal VPS or shared?

**Comment by nikolowry** [comment]
@csagan5 nothing fancy, just a Digital Ocean droplet with 4 vCPUs -- was hoping with a little experimenting and love I could figure out how to optimize for reliable builds. I've never built Chromium on there as of yet, but I'd imagine it will be 4-6 hours.

Does anyone know what /e/ uses for their builds? Their Bromite builds seem to take ~3.5 hours https://gitlab.e.foundation/e/apps/browser/pipelines

**Comment by csagan5** [comment]
> I'd imagine it will be 4-6 hours

I would say more, but let me know.

> Does anyone know what /e/ uses for their builds?

It is not visible from GitLab, they will probably explain if you ask.

**Comment by wchen342** [comment]
I just saw the automatic build for `ungoogled-chromium-archlinux` [timed out](https://github.com/ungoogled-software/ungoogled-chromium-archlinux/pull/55) for v80 because of #928. I think a workaround is possible but it will be a quite ugly one (upload half-finished output as artifact and continue as another workflow).

I think if we got some donations then we can run our own machines, both for building and testing, but for now it seems not to be an option. What does @Eloston think?

**Comment by jstkdng** [comment]
Chromium really needs fast ram more than a lot of cores. 16GB tmpfs should be able to build chromium in 3 hours or so. If you don't have the money just use the OBS.

@csagan5 
You should be able to build in the OBS, as I'm doing here: https://build.opensuse.org/package/show/home:justkidding:arch/ungoogled-chromium
dunno how the android setup would be tho.

**Comment by wchen342** [comment]
@jstkdng I think the goal here is to find a way that can be integrated with github pull requests/tags/etc. Do you know OBS can do that or not?

**Comment by csagan5** [comment]
> You should be able to build in the OBS, as I'm doing here:

@jstkdng thanks for the pointer! I will definitely give a look.

> Do you know OBS can do that or not?

@wchen342 I think you cannot use OBS as slaves; for me the goal is not to necessarily use GitHub actions so OBS will be fine.

**Comment by jstkdng** [comment]
@wchen342 
You could try to use github webhooks as the OBS has its own rest api, but you would have to build all of chromium for every small change.

**Comment by Rowisi** [comment]
> @jstkdng I think the goal here is to find a way that can be integrated with github pull requests/tags/etc. Do you know OBS can do that or not?

No, the goal is to build Chromium just for releases.

> Chromium really needs fast ram more than a lot of cores. 16GB tmpfs should be able to build chromium in 3 hours or so. If you don't have the money just use the OBS.

I can also build Chromium on my computer in less than 3 Hours, but what about others(Windows and MacOS)?

I just want a to find general solution for all platforms, thats the goal of this discussion.

**Comment by wchen342** [comment]
@csagan5 Just a heads up, OBS seems to be prohibiting `chromium` that contains patent codes, see the links in [https://aur.archlinux.org/packages/ungoogled-chromium/#comment-728578](https://aur.archlinux.org/packages/ungoogled-chromium/#comment-728578).

**Comment by Rowisi** [comment]
> @csagan5 Just a heads up, OBS seems to be prohibiting `chromium` that contains patent codes, see the links in https://aur.archlinux.org/packages/ungoogled-chromium/#comment-728578.

I have just downloaded and installed it manually.

**Comment by wchen342** [comment]
> I have just downloaded and installed it manually.

The fact that someone has run into the problem means there is a possibility that the project can get taken down at some time. Maybe the rules are not enforced so well so we can get around it from time to time, but it does not mean we should rely on such loopholes to make things work.

The decision is package maintainer's and I still prefer using Github itself at this time.

**Comment by csagan5** [comment]
> OBS seems to be prohibiting `chromium` that contains patent codes

They must refer to the codecs; in this case then I will not try to build it on OBS.

**Comment by jstkdng** [comment]
@csagan5 
you could just use their own chromium tarball without the patented parts. I'm gonna try doing that: https://build.opensuse.org/source/network:chromium/chromium/chromium-80.0.3987.87.tar.xz

@wchen342 
>The decision is package maintainer's and I still prefer using Github itself at this time.

I'd prefer to use my own hardware tbh, but I don't have enough compute power. Also, the project could also be shut down on github as you are using the chromium tarball with patented stuff as well.

**Comment by wchen342** [comment]
> In our case, it could be in google's best interest to just send a C&D or a DMCA to github.

Technically the file is not saved on Github servers, only exists transiently during build so I'm not sure that's the same thing. Not an expert so I could be wrong on this.

> Edit 2: Just downloaded the tarball from OpenSUSE's repo and the one my repository uses, the sha256 sums are the same.

That's interesting. I think the problem with my Android build and probably `bromite` too is that the usual tarball doesn't include Android codes. I have to check them out using `gclient` manually (I'd like to know how @csagan5 does this actually).

**Comment by jstkdng** [comment]
I'm checking that OpenSUSE blacklist page: https://en.opensuse.org/index.php?title=openSUSE:Build_Service_application_blacklist&action=history
and chromium has been listed there since 2010, maybe in 2010 there **where** patented code portions in the chromium source but now that no longer is the case and they just forgot to update it, so no need to worry.

**Comment by nikolowry** [comment]
@csagan5 here's a brief overview of building release `80.0.3987.78` on my Digital Ocean Droplet with 4 vCPUS, 8gb memory and 165gb storage:

**Duration**: 8h23m
**CPU Usage**: All 4 vCPUs were at maxed out at 100% usage for the entirety of the build
**Memory**: Usage hovered around 2-3gb for the entirety of the build
**Disk Usage**: 45gb

**Comment by Rowisi** [comment]
Thank you @everyone. This discussion has been really interesting. 

I think at the end we need a paid build service to build chromium. No free server would give you the requirements to do such a huge task. Even OBS will stop their free service someday and turn it into paid(I am sure of it, they just want some attention ;) ).

Anyways, you guys know better than me, I am new to this project and I just want this browser to be available for anyone.

**Comment by jstkdng** [comment]
Yeah, the ideal solution would be to have a build server, sort of how arch does (pkgbuild.com) , and to have it build for all distros. We could even install the OBS software into that server and just let it do its thing. Of course, that costs money, like 100+ USD per month (speaking from experience).
Until that money can be obtained every month we just have to wait and use our own solutions for as long as we can.

**Comment by Rowisi** [comment]
@jstkdng Yes, OBS is the best solution atm :)

**Comment by wchen342** [comment]
@jstkdng @Rowisi If we can have a standalone build server then I will definitely vote for that :)

Then reason I want to stick with Github at this time is I want to solve this issue and the unit test issue at once, since Actions can be used as a CI too, and part of the test we need for every pull request is it will build successfully. Otherwise we need to configure two different things runs build at the same time.

**Comment by csagan5** [comment]
@jstkdng OBS does not blacklist Chromium because it can be sued but simply because the build process of Chromium includes binary parts which get embedded in the final binary. This is allowed by the Chromium project also for redistribution, but the pieces which do not come from sources might be patented thus not open source. They stick to OSI licenses only and they should be praised for that.

@wchen342 I do not check anything out with `gclient` manually: I use `gclient sync`, @nikolowry also does something similar. Most of the downloaded binaries are however not used/embedded to build Bromite thanks to the patches to kill such integrations.

> We could even install the OBS software into that server

That would be overkill; if you setup a server for build purposes you can use with GitHub/GitLab directly.

> maybe in 2010 there **were** patented code portions in the chromium source but now that no longer is the case (that would explain same sha256 sums) and they just forgot to update the list

@jstkdng I think someone should ask them directly, considering also that a sizable amount of resources would be needed. The patented bits are still there in Chromium and the criteria for not allowing them are the same F-Droid has for inclusion/exclusion.

> on my Digital Ocean Droplet with 4 vCPUS, 8gb memory and 165gb storage

That would be too expensive and unfortunately not much faster than cheaper options, but thanks for sharing.

**Comment by illiliti** [comment]
OBS doesn't support alpine,void linux(musl libc)
